### PR TITLE
GH-42041: [Swift] Fix nullable type decoder issue

### PR DESCRIPTION
**Pull request by abandy**
### Rationale for this change

There is an issue when decoding nullable types.  The previous method of checking for nil values always returned false for nullable types due too the ArrowArray types being non nullable.

### What changes are included in this PR?
This PR adds a IsNull method to the ArrowDecoder to be used for null checks.  Also, a check for nullable types has been added to the Unkeyed decode method.

### Are these changes tested?
Yes, tests have been added/modified to test this fix.
* GitHub Issue: #42041

### CODE REVIEW

File: swift/Arrow/Sources/Arrow/ArrowDecoder.swift [hunk 1]
```diff
@@ -126,11 +136,19 @@ private struct ArrowUnkeyedDecoding: UnkeyedDecodingContainer {
...
-        if type == Int8.self || type == Int16.self ||
+        if type == Int8?.self || type == Int16?.self ||
+            type == Int32?.self || type == Int64?.self ||
+            type == UInt8?.self || type == UInt16?.self ||
+            type == UInt32?.self || type == UInt64?.self ||
+            type == String?.self || type == Double?.self ||
+            type == Float?.self || type == Date?.self {
+            defer {increment()}
+            return try self.decoder.doDecode(self.currentIndex)!
+        } else if type == Int8.self || type == Int16.self ||
```
Comment (kou) [inline swift/Arrow/Sources/Arrow/ArrowDecoder.swift:151]: I'm not familiar with Swift but can we unify this `if` and `else if`? It seems that body codes of them are same.

File: swift/Arrow/Tests/ArrowTests/CodableTests.swift [hunk 2]
```diff
@@ -165,6 +186,30 @@ final class CodableTests: XCTestCase {
...
+                if index % 2 == 0 {
+                    XCTAssertNil(str!)
+                } else {
+                    XCTAssertEqual(str, "test\(index)")
+                }
+                index += 1
+            }
+        case .failure(let err):
+            throw err
+        }
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/CodableTests.swift:212]: How about creating a new separated test for this?

File: swift/Arrow/Tests/ArrowTests/CodableTests.swift [hunk 3]
```diff
@@ -134,7 +136,26 @@ final class CodableTests: XCTestCase {
...
+            .finish()
+        switch resultWNil {
+        case .success(let rb):
+            let decoder = ArrowDecoder(rb)
+            let testData = try decoder.decode(Int8?.self)
+            for index in 0..<testData.count {
+                let val: Int8? = testData[index]
+                if index % 2 == 1 {
+                    XCTAssertNil(val)
+                } else {
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/CodableTests.swift:158]: How about creating a new separated test for this case?
`testArrowSingleDecoderWithoutNull()` and `testArrowSingleDecoderWithNull()`?

**Review by kou — APPROVED** [review]
+1

File: swift/Arrow/Sources/Arrow/ArrowDecoder.swift [hunk 4]
```diff
@@ -173,7 +191,7 @@ private struct ArrowKeyedDecoding<Key: CodingKey>: KeyedDecodingContainerProtoco
     }
 
     func decodeNil(forKey key: Key) throws -> Bool {
-        return try self.decoder.doDecode(key) == nil
+         try self.decoder.isNull(key)
```
[the remark this answers is not in the document]
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowDecoder.swift:194]: Good catch!  The file is being checked by the linter but apparently a space off doesn't cause it to fail :).  I will update accordingly.

File: swift/Arrow/Sources/Arrow/ArrowDecoder.swift (hunk 1 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowDecoder.swift:151]: They can be unified.  I had separated them to delineate the nullable and non nullable types but I can merge them.

**Review by kou — APPROVED** [review]
+1

**Pull request merged**